src/loading.ts

```diff
@@ -0,0 +1,73 @@
+import { OperatorFunction, merge, combineLatest, of } from 'rxjs'
+import { share, startWith, map, filter, mapTo, delay, endWith, scan } from 'rxjs/operators'
+import { isEqual } from 'lodash'
+
+export const LOADING = 'loading' as const
+
+/**
+ * An emission from a result provider.
+ *
+ * @template T The type of the result. Should typically include an empty value, or even an error type.
+ */
+export interface MaybeLoadingResult<T> {
+    /**
+     * Whether the result provider is currently getting a new result.
+     */
+    isLoading: boolean
+
+    /**
+     * The latest result.
+     */
+    result: T
+}
+
+/**
+ * Maps a stream of MaybeLoadingResult (which contains both results and loading states) to a stream of clear
+ * instructions on when to show a loader, results or nothing.
+ *
+ * @param loaderDelay The delay, in milliseconds, after which a loader should be shown if no results have been emitted.
+ * @param emptyResultValue The value that represents the absence of results. This will be emitted, and also deep-compared to with `isEqual()`. Example: `null`, `[]`
+ *
+ * @template TResult The type of the provider result (without `TEmpty`).
+ * @template TEmpty The type of the empty value, e.g. `null` or `[]`.
+ */
+export const emitLoading = <TResult, TEmpty>(
+    loaderDelay: number,
+    emptyResultValue: TEmpty
+): OperatorFunction<MaybeLoadingResult<TResult | TEmpty>, TResult | TEmpty | typeof LOADING | undefined> => source => {
+    const sharedSource = source.pipe(
+        // Prevent a loading indicator to be shown forever if the source completes without a result.
+        endWith<Partial<MaybeLoadingResult<TResult | TEmpty>>>({ isLoading: false }),
```

Comment:
nice catch

Reply:
Your comment actually made me realize there are no tests for this case, so I added more tests and realized there was a slight bug with this edge case (loader delay would still block completion). Fixed in https://github.com/sourcegraph/codeintellify/pull/251/commits/856007bb2b4109cedad1ef4f4d14852e583fc200, thanks 😊